HDDS-11509. logging improvements for directory deletion

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?

All Ozone components involved in background deletion should follow consistent patterns with their logging:
At the end of each run, and INFO log should display how many items were processed, what the current limit of items per run is, and how long it took.
After each run, a DEBUG log should indicate each individual item that was processed.

OM DirectoryDeletingService:
- [x] INFO log for no. of items deleted ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java#L463C7-L467C36))
- [x] INFO log for time taken by one iteration ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java#L463C7-L467C36))
- [x] INFO log for limit per iteration (added)
- [x] WARN log when iteration takes longer than configured ([exisiting](https://github.com/apache/ozone/blob/efe589297c471100c5348f57fdf8aa91d8968795/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/BackgroundService.java#L151C13-L154C14))
- [x] DEBUG log of each item processed ([existing](https://github.com/apache/ozone/blob/efe589297c471100c5348f57fdf8aa91d8968795/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java#L349C4-L352C6))

OM KeyDeletingService:
- [x] INFO log for no. of items deleted ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java#L118))
- [x] INFO log for time taken by one iteration ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java#L118))
- [x] INFO log for limit per iteration (added)
- [x] WARN log when iteration takes longer than configured ([exisiting](https://github.com/apache/ozone/blob/efe589297c471100c5348f57fdf8aa91d8968795/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/BackgroundService.java#L151C13-L154C14))
- [x] DEBUG log of each item processed ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java#L106C5-L109C5))

OM OpenKeyCleanUpService:
- [x] INFO log for no. of items deleted (added)
- [x] INFO log for time taken by one iteration (added)
- [x] INFO log for limit per iteration (added)
- [x] WARN log when iteration takes longer than configured ([existing](https://github.com/apache/ozone/blob/efe589297c471100c5348f57fdf8aa91d8968795/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/BackgroundService.java#L151C13-L154C14))
- [x] DEBUG log of each item processed (added)

DN BlockDeletingService:
- [x] INFO log for no. of items deleted ([existing](https://github.com/Tejaskriya/ozone/blob/16ba289fe32cfafe1b1b6e0ef31c6cd349e3c0af/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/interfaces/ContainerDeletionChoosingPolicyTemplate.java#L85C4-L86C76))
- [x] INFO log for time taken by one iteration (added)
- [x] INFO log for limit per iteration (added)
- [x] WARN log when iteration takes longer than configured ([existing](https://github.com/apache/ozone/blob/efe589297c471100c5348f57fdf8aa91d8968795/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/BackgroundService.java#L151C13-L154C14))
- [x] DEBUG log of each item processed (added for each container)

SCM SCMBlockDeletingService:
- [x] INFO log for no. of items deleted ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/SCMBlockDeletingService.java#L205))
- [x] INFO log for time taken by one iteration([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/SCMBlockDeletingService.java#L206))
- [x] INFO log for limit per iteration (added)
- [x] WARN log when iteration takes longer than configured ([existing](https://github.com/apache/ozone/blob/efe589297c471100c5348f57fdf8aa91d8968795/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/BackgroundService.java#L151C13-L154C14))
- [x] DEBUG log of each item processed ([existing](https://github.com/apache/ozone/blob/624bede619004a6a2be5fe185abd40c2999ae3c1/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/SCMBlockDeletingService.java#L196))

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11509

## How was this patch tested?

Manually tested

aryangupta1998 commented:
Thanks, @Tejaskriya for the patch. Regarding the master [jira](https://issues.apache.org/jira/browse/HDDS-11506), I anticipate significant changes to the deletion code. So, I think we should address the logging improvement JIRA later to make sure the logging can be implemented more effectively.

errose28 reviewed (commented):
Thanks for working on this @Tejaskriya. I think we should move forward with this even if some apsects may change later as part of the parent Jira. There is no timeline for most of the delete changes mentioned there so this will provide us help in the mean time.

To make this easier to follow, can you add a checklist to the PR description to track what logs we have for each service? It can look something like this (just an example, the actual items aren't correct):
- OM Open Key Cleanup Service:
  - [x] DEBUG log of each item processed (added)
  - [x] INFO log with summary of the last run [existing](https://github.com/apache/ozone/blob/0f1619573036090b772efa3f6db8fcde7d1ccc5c/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/OpenKeyCleanupService.java#L232-L233)
  - [x] WARN log when limits are hit (added)
- OM key deleting service
...

The items can be checked off as they are covered in the PR. Each item can indicate whether it was added in this PR, or provide a github link to where it is already present in the code.

Tejaskriya commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/BlockDeletingService.java:
```diff
@@ -153,8 +155,8 @@ public BackgroundTaskQueue getTasks() {
       if (containers.size() > 0) {
-        LOG.debug("Queued {} blocks from {} containers for deletion",
-            totalBlocks, containers.size());
+        LOG.info("Queued {} blocks from {} containers for deletion, blocksLimit was {}, elapsed time {}ms.",
+            totalBlocks, containers.size(), blocksLimitPerInterval, Time.monotonicNow() - startTime);
```
Do we simply convert the debug log you have pointed out to an info log, or do we want to add a info log in BlockDeletingService pointing out each container + number of blocks deleted for that container when it is added the the queue?

Tejaskriya commented:
@ashishkumar50 @aryangupta1998 Could you please review the patch when you get some time?

aryangupta1998 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:
```diff
@@ -449,13 +453,21 @@ public long optimizeDirDeletesAndSubmitRequest(long remainNum,
+            OMConfigKeys.OZONE_PATH_DELETING_LIMIT_PER_TASK, ozoneManager.getConfiguration()
+                .getInt(OMConfigKeys.OZONE_PATH_DELETING_LIMIT_PER_TASK,
+                    OMConfigKeys.OZONE_PATH_DELETING_LIMIT_PER_TASK_DEFAULT));
+      }
```
This shouldn't be a warn log, we can add it in a debug log and simplify it,

> if (remainNum <= 0 && LOG.isDebugEnabled()) {
      LOG.debug(
          "Limit for number of directories and files that can be processed in a deletion interval is reached. Total files and dirs moved: {}",
          (subFileNum + subDirNum - subdirDelNum));
    }

errose28 replied:
I guess the Jira description doesn't currently specify what do log when the deletion limits are hit:
> At the end of each run, and INFO log should display how many items were processed, what the current limit of items per run is, and how long it took.
If the run time of the iteration exceeded the configured interval, the message should highlight this and be elevated to WARN.
This should also log the config keys that can remedy this for batch size, interval, or number of threads.

However based on past experiences troubleshooting slow deletion issues in a busy cluster with many other events going on I would be in favor of also elevating this to warn if the limits are being hit to indicate that the system is under load and deletion may be under-provisioned. However from the Jira this should be one single message as a summary with all this information at the corresponding log level based on its content.

Also worth noting that once the multi-threaded implementation is complete it should take a lot more to hit the deletion limit, and this message will not seem so spurious if 10 threads are being maxed out trying to process current deletion.

aryangupta1998 replied:
However, with a million directories, the log could become flooded with this warning, but we can keep it for now.

errose28 replied:
The number of log lines won't change, it will be one per iteration regardless of the conditions. We should definitely warn if it is taking longer than the interval. As for the key count limits being hit, it could go either way. One option is we could leave that condition at info for now, and as a follow up we could optionally save the state of the last few runs and only log a warning if the last 10 or so hit the key limits.

aryangupta1998 replied:
Yes, that's a nice idea. We can warn if we hit the key limit more than 10 times.

aryangupta1998 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/DirectoryDeletingService.java:
```diff
@@ -154,11 +154,11 @@ public int getPriority() {
+        final long run = getRunCount().incrementAndGet();
         if (LOG.isDebugEnabled()) {
-          LOG.debug("Running DirectoryDeletingService");
+          LOG.debug("Running DirectoryDeletingService {}", run);
```
We already log the run count in optimizeDirDeletesAndSubmitRequest() present inAbstractKeyDeletingService.java, do we need it here as well?

Tejaskriya replied:
In order to keep the log consistent across the various deleting services, I have added this. We can easily identify each run of every deleting service instead of having to search in the logs.

aryangupta1998 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:
```diff
@@ -449,13 +453,21 @@ public long optimizeDirDeletesAndSubmitRequest(long remainNum,
+          timeTakenInIteration, getRunCount());
+      if (remainNum <= 0) {
+        LOG.warn("Limit for no. of directory+files that can be deleted in one iteration is reached. " +
+                "Current limit: {} = {}",
```
IIUC, this can be a little confusing, like, 
`Current limit: 10000 = 6000`
Let's only display the current limit with the configured value.

Tejaskriya replied:
I have removed the warning logs now.

sumitagrawl reviewed (commented):
Please find few comments...

Tejaskriya commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/BlockDeletingService.java:
```diff
@@ -149,12 +151,19 @@ public BackgroundTaskQueue getTasks() {
         containerBlockInfos = builder.build();
         queue.add(containerBlockInfos);
         totalBlocks += containerBlockInfo.getNumBlocksToDelete();
+        LOG.info("Queued- Container: {}, deleted blocks: {}",
```
I have changed it to debug level now

Tejaskriya commented:
Thanks for the reviews @aryangupta1998 @sumitagrawl. I have removed the warn logs and addressed the other comments. Could you please take another look?

sumitagrawl commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/BlockDeletingService.java:
```diff
@@ -149,12 +151,14 @@ public BackgroundTaskQueue getTasks() {
       if (containers.size() > 0) {
-        LOG.debug("Queued {} blocks from {} containers for deletion",
-            totalBlocks, containers.size());
+        LOG.info("In this iteration, blocksLimit was {}, elapsed time {}ms.", blocksLimitPerInterval,
```
already have below log in methods, so this is redundant log, do not need to have this.
```
LOG.info("Chosen {}/{} blocks from {} candidate containers.",
        (originalBlockCount - blockCount), blockCount, orderedList.size());
```

Tejaskriya replied:
The log you have mentioned prints the blocks details, the time taken by the iteration and the limit is not showed. It would be good to have it uniformly present in all the deletion services logs.

sumitagrawl replied:
> The log you have mentioned prints the blocks details, the time taken by the iteration and the limit is not showed. It would be good to have it uniformly present in all the deletion services logs.

Add to the other log if required instead of adding new logs.

Tejaskriya commented:
Thanks for the review @sumitagrawl, I have addressed the comments!

sumitagrawl reviewed (approved):
LGTM
